errors/errors.go

```diff
@@ -0,0 +1,195 @@
+package errors
+
+import (
+	"fmt"
+
+	"github.com/pkg/errors"
+)
+
+var (
+	// InternalErr represents a general case issue that cannot be
+	// categorized as any of the below cases.
+	InternalErr = NewRootError(0, "internal")
+
+	// UnauthorizedErr is used whenever a request without sufficient
+	// authorization is handled.
+	UnauthorizedErr = NewRootError(1, "unauthorized")
+
+	// NotFoundErr is used when a requested operation cannot be completed
+	// due to missing data.
+	NotFoundErr = NewRootError(2, "not found")
+
+	// InvalidMsgErr is returned whenever an event is invalid and cannot be
+	// handled.
+	InvalidMsgErr = NewRootError(3, "invalid message")
+
+	// InvalidModelErr is returned whenever a message is invalid and cannot
+	// be used (ie. persisted).
+	InvalidModelErr = NewRootError(4, "invalid model")
+)
+
+// NewRootError returns an error instance that should be used as the base for
+// creating error instances during runtime.
+//
+// Popular root errors are declared in this package, but extensions may want to
+// declare custom codes. This function ensures that no error code is used
+// twice. Attempt to reuse an error code results in panic.
+//
+// Use this function only during a program startup phase.
+func NewRootError(code uint32, description string) RootError {
+	if e, ok := usedCodes[code]; ok {
+		panic(fmt.Sprintf("error with code %d is already registered: %q", code, e.desc))
```

Comment:
q: to avoid the panic we could return the previous registered error as well. WDYT?

Reply:
I think this is a fatal mistake. Returning previous error could lead to a lot of confusion, for example:
```go
// in x/foo/foo.go
ConflictErr = NewRootError(4, "not unique")

// in x/bar/bar.go
ConstErr = NewRootError(4, "constant")
```
Depending on the compilation order, only one error will be assigned to both. But then, during the runtime, you will get a missleading error assuming that `foo.ConflictErr` and `bar.ConstErr` are dirrerent.

What I would like instead is this to be fixed before committing the code :man_teacher:

Reply:
It would be definitively confusing to return a different message. If the function is used proper during initialization phase I don't see a problem, only at runtime. 
In https://github.com/iov-one/weave/pull/295/files#diff-67bcaf0cce945ce4f52191db72018763R175 equality is defined by the `code` (when not internal) so I was wondering whey it must be different here. :-)

Reply:
I am not sure if I follow.

I think it is a mistake if you use the same code (number) more than once. This can happen because you can register top-level errors in extensions as well. Definitions are split between files in different directories.

Below is a valid Go code, but an invalid logic:
```go
package myextension

var (
    ErrOne = errors.Register(1, "one")
    ErrFirst = errors.Register(1, "first")
)
```
There must not be more than one error with code 1. How to deal with such situation?

Reply:
@husio global registry, panic on trying to register something that has been registered already, fail fast and make user fix the issue.

Reply:
@ruseinov This is how it works right now. Alex is considering a different approach and this is what we are discussing here.

Reply:
I am 🆗 with this `panic`. It is likely that we notice it during development when we register the error during var initialization.